File: feed-reader/src/main/scala/name/abhijitsarkar/scala/Transformer.scala
package name.abhijitsarkar.scala

import java.nio.file.{DirectoryStream, Path}

import akka.actor.{Actor, Props}
import akka.stream.scaladsl.Source

import scala.collection.immutable.{List => ImmutableList}

/**
  * @author Abhijit Sarkar
  */

case class Message(dir: DirectoryStream[Path], text: String)

import scala.collection.JavaConverters._

class Transformer extends Actor {
  def receive = {
    case Message(dir, text) => {
      val flow = Source(dir.asScala.toList).map(p => {
        val lines = io.Source.fromFile(p.toFile).getLines().filter(_.contains(text)).map(_.trim).to[ImmutableList]
        (p.toAbsolutePath.toString, lines)
      }).filter(!_._2.isEmpty)

      sender ! flow
    }
  }
}

object Transformer {
  def props = Props(new Transformer)
}

File: feed-reader/src/main/java/name/abhijitsarkar/java/Transformer.java
package name.abhijitsarkar.java;

import rx.Observable;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.DirectoryStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Collection;
import java.util.List;
import java.util.Map;

import static java.util.stream.Collectors.collectingAndThen;
import static java.util.stream.Collectors.groupingBy;

/**
 * @author Abhijit Sarkar
 */
public class Transformer {
    public Observable<Map.Entry<String, ? extends Collection<String>>> transform(DirectoryStream<Path> dir, String text) {
        return Observable.from(dir).flatMap(p -> {
            try {
                return Observable.from((Iterable<Map.Entry<String, List<String>>>) Files.lines(p)
                        .filter(line -> line.contains(text))
                        .map(String::trim)
                        .collect(collectingAndThen(groupingBy(pp -> p.toAbsolutePath().toString()), Map::entrySet)));
            } catch (IOException e) {
                throw new UncheckedIOException(e);
            }
        });
    }
}
